content/blog/2019/07/2019-07-15-pipeline-config-history-plugin.adoc

```diff
@@ -0,0 +1,88 @@
+It is also possible to recover old configuration versions.
+Today the plugin has more than 30,000 installations.
+For many years JobConfigHistory relieves our daily work -- with more than 3,000 Jenkins jobs!
+Then there was a new type of job: **Pipelines**.
+
+
+== Pipelines - something new was needed
+
+Pipeline jobs differ from classical job types fundamentally.
+While classic job types get configured via the Jenkins GUI, Pipeline jobs get configured as code.
+Every pipeline job indeed gets created via the Jenkins GUI, however that is not necessarily where the pipeline configuration is located.
+There are different ways where pipelines can get configured:
+
+* Directly in the Jenkins job as script.
+  The code gets inserted directly in the job configuration page.
+* As Jenkinsfile in the source code management system (SCM): The pipeline configuration is defined in a text file (Jenkinsfile) in the SCM.
+  In the job itself only the path to the repository of the Jenkinsfile is configured.
+  During the build the Jenkinsfile gets checked out from the SCM and processed.
+* As shared library: A part of the pipeline configuration gets moved to own files, that can be used by several jobs.
+  These files are also saved in the SCM.
+  Even so a Jenkinsfile is still needed (or a pipeline script in the job).
+
+With every save operation of the job configuration, JobConfigHistory creates a copy of the actual job configuration if something has changed.
+For pipeline jobs that only works if the pipeline configuration is inserted in the job configuration page as script.
+Jenkinsfiles or shared libraries cannot get detected by JobConfigHistory.
+You have to go to the SCM to view changes of the Jenkinsfiles or shared libraries.
+It is complex and time intensive to find a correlation between the time of a build and a change to the Jenkinsfile or shared library.
+
+This new problem cannot get solved by JobConfigHistory and a new solution were needed to get found, that also detects these changes and provides a way to view these changes within the Jenkins GUI.
+So we developed Pipeline Configuration History.
+
+During every pipeline run the Jenkinsfile and used shared libraries get saved in the `builds` directory of the job.
+The function of Pipeline Configuration History is to check after every pipeline run, if there were changes of the pipeline files between the last and the previous run and if so, it saves these files as history events.
+Therefore when a pipeline job ceases to build successfully, you can check if something has changed on any used pipeline file.
+You can also see the build where changes occurred.
+
+image::/images/post-images/2019-07-pipeline-config-history/image2019-5-15_13-44-54.png[Screenshot of Pipeline Configuration History]
+
+Because a pipeline configuration can consist of several files where changes could have occured, only those files in which something has changed between two builds get shown at the diff.
```

Comment:
```suggestion
Because a pipeline configuration can consist of several files where changes could have occurred, only files with changes between two builds are shown in the diff.
```